dra-evolution/pkg/api/capacity_types.go

```diff
@@ -3,168 +3,134 @@ package api
-	// block size should be "4Ki". Default is 1.
+// DeviceAttribute is a combination of an attribute name and its value.
+// Exactly one value must be set.
+type DeviceAttribute struct {
+	// Name is a unique identifier for this attribute, which will be
+	// referenced when selecting devices.
 	//
-	// If the resource is consumable in a fractional way, then the
-	// default of 1 should not be used; instead this should be a fractional
-	// amount corresponding the increment size. We may also need a minimum
-	// value, if the minimum is larger than the block size (as is the case
-	// for standard Kubernetes CPU resources).
+	// Attributes are defined either by the owner of the specific driver
+	// (usually the vendor) or by some 3rd party (e.g. the Kubernetes
+	// project). Because attributes are sometimes compared across devices,
+	// a given name is expected to mean the same thing and have the same
+	// type on all devices.
 	//
-	// +optional
-	BlockSize *resource.Quantity `json:"blockSize,omitempty"`
-}
-
-// Attribute capture the name, value, and type of an device attribute.
-type Attribute struct {
-	Name string `json:"name"`
-
-	// One of the following:
-	StringValue   *string            `json:"stringValue,omitempty"`
-	IntValue      *int               `json:"intValue,omitempty"`
-	QuantityValue *resource.Quantity `json:"quantityValue,omitempty"`
-	SemVerValue   *SemVer            `json:"semVerValue,omitempty"`
-}
-
-func (a Attribute) Equal(b Attribute) bool {
-	if a.Name != b.Name {
-		return false
-	}
-
-	return a.EqualValue(b)
-}
-
-func (a Attribute) EqualValue(b Attribute) bool {
-	if a.StringValue != nil && b.StringValue != nil && *a.StringValue == *b.StringValue {
-		return true
-	}
-
-	if a.IntValue != nil && b.IntValue != nil && *a.IntValue == *b.IntValue {
-		return true
-	}
-
-	if a.QuantityValue != nil && b.QuantityValue != nil && (*a.QuantityValue).Equal(*b.QuantityValue) {
-		return true
-	}
-
-	if a.SemVerValue != nil && b.SemVerValue != nil && *a.SemVerValue == *b.SemVerValue {
-		return true
-	}
-
-	return false
+	// Attribute names must be either a C-style identifier
+	// (e.g. "the_name") or a DNS subdomain followed by a slash ("/")
+	// followed by a C-style identifier
+	// (e.g. "example.com/the_name"). Attributes whose name does not
+	// include the domain prefix are assumed to be part of the driver's
+	// domain. Attributes defined by 3rd parties must include the domain
+	// prefix.
+	//
+	// The maximum length for the DNS subdomain is 63 characters (same as
+	// for driver names) and the maximum length of the C-style identifier
+	// is 32.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// The Go field names below have a Value suffix to avoid a conflict between the
+	// field "String" and the corresponding method. That method is required.
+	// The Kubernetes API is defined without that suffix to keep it more natural.
+
+	// QuantityValue is a quantity.
```

Comment:
Candidate for list-of-one-of pattern because we might add more types

Suppose we add `FloatValue`.

A forward-rev driver which tries to set a float will fail at the (back-rev) API because none of the known fields are set.  Good.

A back-rev scheduler will know something is wrong because none of the known fields are set.  Good.

So this seems OK?

Reply:
I agree. Embedded one-of vs. "struct with some shared fields (name here in this case) and some one-of fields (the value fields here)" have the same semantic, as long as clients are aware.